### PR TITLE
samples: nrf9160: modem_shell: simple REST client

### PR DESCRIPTION
**Pull request by jhirsi**
New shell command for sending and receiving simple REST queries and responses.
Uses `REST_CLIENT` library.

----------------------------------------------

Usage examples:
HEAD:
```
mosh:~$ rest -d example.com -l 1024 -m head
RRC mode: Connected
Response:
  Socket ID: 4 (sckt closed: yes)
  HTTP status: 200
  Body/Total length: 0/332


[00:33:42.869,201] <wrn> rest_client: No data in a response body
```
HTTPS POST for generating a nRF Cloud service evaluation token:

```
mosh:~$ rest -m post -p 443 -s 16842753 -d "api.nrfcloud.com" -u "/v1/account/service-evaluation-token" -H "Authorization: Bearer <nRF Cloud API Key>\x0D\x0A" -H "Content-Type: application/json\x0D\x0A"
RRC mode: Connected
Response:
  Socket ID: 4 (sckt closed: yes)
  HTTP status: 200
  Body/Total length: 316/551

  {"token":"<returned service evaluation token>","expires":"2021-11-28T08:45:14.942Z"}

mosh:~$

```

POST for requesting a location based on cellular parameters (using generated service evaluation token):

```
mosh:~$ rest -m post -p 443 -s 16842753 -d "api.nrfcloud.com" -u "/v1/location/cell" -H "Authorization: Bearer <service evaluation token>\x0D\x0A" -H "Content-Type: application/json\x0D\x0A" --body '{"lte":[{"eci":19715,"mcc":244,"mnc":12,"tac":2002,"earfcn":1598,"rsrp":-110,"nmr":[{"earfcn":1598,"pci":250,"rsrp":-117,"rsrq":-3},{"earfcn":300,"pci":227,"rsrp":-119}]}]}\x0D'
RRC mode: Connected
Response:
  Socket ID: 4 (sckt closed: yes)
  HTTP status: 200
  Body/Total length: 79/313

  {"lat":61.464941,"lon":23.810654,"uncertainty":92,"fulfilledWith":"MCELL_EVAL"}

```

### CODE REVIEW

File: samples/nrf9160/modem_shell/src/srest/srest_shell.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef SREST_RIP_SHELL_H
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.h:7]: Yeah, that (REST In Peace) was my original name for this command ;-)
Will fix this.

> A generic comment about naming, does the name have to include the word "simple"? It could just be "REST client" and the command "rest", IMHO.
I wanted to emphasize that this only supports only SIMPLE queries/responses. None kind of streaming and whole response need to fit in same buffer, just a request and a response...

[the remark this answers is not in the document]
Comment (tokangas) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.h:7]: Ok, keeping "simple" is fine by me.

[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.h:7]: I guess "simple" is fine although there is a chance this is developed to a more advanced version at a later stage.

File: samples/nrf9160/modem_shell/Kconfig [hunk 2]
```diff
@@ -70,6 +70,13 @@ config MOSH_FOTA
 	help
 	  FOTA for performing software updates over-the-air for both modem and application side.
 
+config MOSH_SIMPLE_REST_CLIENT
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/Kconfig:73]: I would probably rename this based on directory and file names, i.e., MOSH_SREST_CLIENT. Just like most of our directories.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+#include <net/rest_client.h>
+
+#include "srest_shell.h"
+#include "mosh_print.h"
+
+static const char srest_shell_cmd_usage_str[] =
+	"Simple REST shell client\n"
+	"Usage:\n"
+	"srest -d host [-p port] [-m method] [-u url] [-b body] [-H header1] [-H header2 ... header10]\n"
+	"               [-s sec_tag] [-v verify_level] [-t timeout_in_ms] [-l length of the response buffer]\n"
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/srest/srest_shell.c:24]: `length_of_the_response_buffer`, or maybe `response_buffer_length` for -l option?

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+					headers_set = true;
+					strcpy(headers[i].header_str, optarg);
+					break;
+				}
+			}
+			break;
+		}
+		case 'v':
+			req_ctx.tls_peer_verify = atoi(optarg);
+			if (req_ctx.tls_peer_verify  < 0 || req_ctx.tls_peer_verify  > 2) {
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/srest/srest_shell.c:228]:
```suggestion
			if (req_ctx.tls_peer_verify < 0 || req_ctx.tls_peer_verify > 2) {
```

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+			for (j = 0; j < SREST_REQUEST_MAX_HEADERS; j++) {
+				headers_set = false;
+				if (headers[j].in_use) {
+					headers_set = true;
+					headers[j].in_use = false;
+					break;
+				}
+			}
+			if (headers_set) {
+				req_headers[i] = headers[j].header_str;
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/srest/srest_shell.c:262]: I'm likely missing something important but I want to raise the question anyway. What is the reason why we are using `headers[j].header_str` as an intermediate storage for the header, i.e., why we don't put it directly into `req_headers[i]` in lines 209-220?

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+			if (req_ctx.port == 0) {
+				mosh_warn("port not an integer (> 0)");
+				return -EINVAL;
+			}
+			break;
+		case 'b':
+			req_ctx.body = optarg;
+			len = strlen(optarg);
+			if (len > 0) {
+				req_ctx.body = k_calloc(len + 1, 1);
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c:178]: I'm putting my comment here too. All places where we return from this function, there is potentially memory leak and not only for this body but also other dynamic memory (headers and req_ctx.resp_buff).

File: samples/nrf9160/modem_shell/src/srest/srest_shell.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.h:7]: Having gone through the code, I'd probably think of renaming to just `rest` instead of `srest` and Simple REST to be more future proof.

File: samples/nrf9160/modem_shell/CMakeLists.txt [hunk 7]
```diff
@@ -29,3 +29,4 @@ add_subdirectory_ifdef(CONFIG_MOSH_GNSS src/gnss)
 add_subdirectory_ifdef(CONFIG_MOSH_SMS src/sms)
 add_subdirectory_ifdef(CONFIG_MOSH_PPP src/ppp)
 add_subdirectory_ifdef(CONFIG_MOSH_FOTA src/fota)
+add_subdirectory_ifdef(CONFIG_MOSH_SIMPLE_REST_CLIENT src/srest)
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/CMakeLists.txt:32]: Not related to this line at all but I wanted a conversation to track.

README.rst should be updated to include new feature/command.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c:262]: A good comment. I'm using intermediate storage for helping the logics, especially when multiple headers. That was my 1st write for this code and most probably this can be done in easier ways. Anyways, I'll simplify the logics based on other Tommi's comments.

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c:262]: Yep, I thought this intermediate storage was coming from the first draft implementation.
I think the code would simplify a lot if you get rid of `headers` and `struct rip_header_list_item`, and use NULL to indicate a free slot in `req_headers`.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.h:7]: OK, let the `srest` rest in peace and let there be a command `rest`.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c:178]: yep, a good catch. Will fix.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+	if (!host_set) {
+		mosh_error("Please, give all mandatory options");
+		goto show_usage;
+	}
+
+	req_ctx.header_fields = NULL;
+	if (headers_set) {
+		/* Fill given headers to req_headers[] from headers[]: */
+		for (i = 0; i < SREST_REQUEST_MAX_HEADERS; i++) {
+			for (j = 0; j < SREST_REQUEST_MAX_HEADERS; j++) {
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: True. Simplified per other Tommi's comments and no need at all for this.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+	"                      -H \"Content-Type: application/json\\x0D\\x0A\"\n"
+	"  -b, --body,         Payload body, example: -b '{\"foo\":bar}'\n"
+	"  -s, --sec_tag,      Used security tag for TLS connection\n"
+	"  -t, --timeout,      Request timeout in seconds\n"
+	"                      (default: CONFIG_REST_CLIENT_REST_REQUEST_TIMEOUT)\n"
+	"  -k, --keepalive,    Keep socket connection alive for upcoming requests\n"
+	"                      towards the ame server (default: false)\n"
+	"  -i, --sckt_id,      Use existing socket id (default: create a new connection)\n"
+	"  -v, --peer_verify,  TLS peer verification level. None (0),\n"
+	"                      optional (1) or required (2). Default value is 2.\n";
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Changed

File: samples/nrf9160/modem_shell/Kconfig (hunk 2 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/Kconfig]: Done.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 4 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Done

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Excellent comment. Thanks @trantanen !

Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Simplified.

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 3 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Done

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: ..and using k_malloc

File: samples/nrf9160/modem_shell/src/srest/srest_shell.c [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+				if (headers[j].in_use) {
+					headers_set = true;
+					headers[j].in_use = false;
+					break;
+				}
+			}
+			if (headers_set) {
+				req_headers[i] = headers[j].header_str;
+				req_headers[i + 1] = NULL;
+				//mosh_print("header %i set: %s", i, req_headers[i]);
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/srest/srest_shell.c]: Yes, was left for debugging. Removed.

File: samples/nrf9160/modem_shell/CMakeLists.txt (hunk 7 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/CMakeLists.txt]: Added a section about new `rest` command. Thus, welcome @peknis to review.

File: samples/nrf9160/modem_shell/README.rst [hunk 11]
```diff
@@ -441,6 +441,26 @@ Examples
...
      ppp uartconf --baudrate 460800
 
+----
+
+REST client
+===========
+
+MoSh command: ``rest``
+
+You can use the REST client for sending and receiving simple REST requests and responses.
```
Comment (peknis) [inline samples/nrf9160/modem_shell/README.rst:451]:
```suggestion
You can use the REST client for sending simple REST requests and receiving responses to them.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/nrf9160/modem_shell/README.rst (hunk 11 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/README.rst]: Done

File: samples/nrf9160/modem_shell/Kconfig [hunk 12]
```diff
@@ -70,6 +70,13 @@ config MOSH_FOTA
 	help
 	  FOTA for performing software updates over-the-air for both modem and application side.
 
+config MOSH_REST_CLIENT
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/Kconfig:73]: Should we shorten this to the command name as for other commands so it would be just `MOSH_REST`
Sorry, I didn't realize this in my earlier comments while focusing still on the "Simple REST" concept.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+#include <getopt.h>
+
+#include <net/http_parser.h>
+#include <net/rest_client.h>
+
+#include "rest_shell.h"
+#include "mosh_print.h"
+
+static const char rest_shell_cmd_usage_str[] =
+	"Simple REST shell client\n"
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/rest/rest_shell.c:21]: Maybe removing word "simple" from here. It's not wrong and, e.g., in documentation I guess it could still be used in few places but I'd probably remove it from the descriptions in the code.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+	mosh_print_no_format(rest_shell_cmd_usage_str);
+}
+
+/*****************************************************************************/
+
+#define REST_REQUEST_MAX_HEADERS 10
+#define REST_RESPONSE_BUFF_SIZE 1024
+#define REST_DEFAULT_DESTINATION_PORT 80
+
+struct rest_header_list_item {
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/rest/rest_shell.c:75]: I guess this is not used anymore and can be removed.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+int rest_shell(const struct shell *shell, size_t argc, char **argv)
+{
+	int opt, i, len;
+	int ret = 0;
+	int long_index = 0;
+	bool show_usage = false;
+	struct rest_client_req_context req_ctx = { 0 };
+	struct rest_client_resp_context resp_ctx = { 0 };
+	/* Headers to request + 1 for NULL: */
+	char *req_headers[REST_REQUEST_MAX_HEADERS + 1] = { NULL };
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/rest/rest_shell.c:89]: Do we really need +1 in the size? The reason I ask this is that the use of `req_headers` seems to always scroll through the table with condition `i < REST_REQUEST_MAX_HEADERS` and I couldn't find where we actually access the element `req_headers[REST_REQUEST_MAX_HEADERS]`.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+#include <net/rest_client.h>
+
+#include "rest_shell.h"
+#include "mosh_print.h"
+
+static const char rest_shell_cmd_usage_str[] =
+	"Simple REST shell client\n"
+	"Usage:\n"
+	"rest -d host [-m method] [-p port] [-s sec_tag] [-v verify_level] [-u url] [-t timeout_in_ms]\n"
+	"               [-H header1] [-H header2 ... header10] [-b body] [-l response_buffer_length]\n"
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/rest/rest_shell.c:24]: `[-H header2 ... header10]`
This gives the impression that you can put several headers for an argument of one -H option. Is that the intention? If not, I would probably just put (without knowing if there is really a common practice to indicate how many times a parameter can be given):
`[-H header1] [-H header2] ... [-H header10]`

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 13 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:21]: You really don't like it as simple ;-)
OK, I can remove it but this really is a SIMPLE REST client.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 15 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:89]: It is needed as the comment above says for ending NULL (it goes all the way to http_client) and we are initializing it here as NULL and intentionally don't touch it elsewhere.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 16 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:24]: Yes, that was NOT an intention. A good point, will change that.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 13 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:21]: Yep, but I still left it in the documentation :)

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 15 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c:89]: OK, so http_client lib needs the final NULL element. Please mention http_lib in the comment so it's very clear and nobody removes it in the future.

File: samples/nrf9160/modem_shell/Kconfig (hunk 12 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/Kconfig]: Changed.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 14 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c]: True, removed.

File: samples/nrf9160/modem_shell/src/rest/rest_shell.c (hunk 13 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c]: yeah, and stays until it is something else than simple ;-)

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/rest/rest_shell.c]: That's fine. When you make this more advanced, there is just one "simple" word in the documentation to be removed :)

**Review by trantanen — APPROVED** [review]

**Review by tokangas — APPROVED** [review]

**Review by hiltunent — APPROVED** [review]

**Comment by peknis** [comment]
We need a changelog entry.

**Comment by jhirsi** [comment]
> We need a changelog entry.

Added one. @peknis please, check.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 17]
```diff
@@ -137,6 +137,10 @@ nRF9160 samples
 
   * Modified to use runtime location service selection instead of compile-time configurations.
 
+* :ref:`modem_shell` sample:
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:140]:
```suggestion
* :ref:`modem_shell_application` sample:
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 17 above)
Comment (jhirsi) [reply doc/nrf/releases/release-notes-changelog.rst]: Done

**Review by peknis — APPROVED** [review]
